File: bindings/zig/ove/src/net_httpd.zig
// Copyright (C) 2026 Kamil Lulko <kamil.lulko@gmail.com>
//
// SPDX-License-Identifier: GPL-3.0-or-later
//
// This file is part of oveRTOS.

//! Embedded HTTP server (singleton).
//!
//! Routes are registered with typed Zig handler functions via a comptime
//! trampoline — no raw C function pointers in application code.

const std = @import("std");
const c = @import("c.zig").raw;
const err = @import("error.zig");
const Error = err.Error;

// ---------------------------------------------------------------------------
// Request / ResponseWriter
// ---------------------------------------------------------------------------

/// HTTP request (opaque, borrowed for handler duration).
pub const Request = struct {
    raw: *c.ove_httpd_req_t,

    /// HTTP method string ("GET", "POST", ...).
    pub fn method(self: Request) [*:0]const u8 {
        return @ptrCast(c.ove_httpd_req_method(self.raw));
    }

    /// Full request path.
    pub fn path(self: Request) [*:0]const u8 {
        return @ptrCast(c.ove_httpd_req_path(self.raw));
    }

    /// Query string or null.
    pub fn query(self: Request) ?[*:0]const u8 {
        const p = c.ove_httpd_req_query(self.raw);
        if (p == null) return null;
        return @ptrCast(p);
    }

    /// POST body as a byte slice (empty if none).
    pub fn body(self: Request) []const u8 {
        const p = c.ove_httpd_req_body(self.raw);
        const len = c.ove_httpd_req_body_len(self.raw);
        if (p == null or len == 0) return &.{};
        const ptr: [*]const u8 = @ptrCast(p);
        return ptr[0..len];
    }

    /// Path segment by index (e.g. "/api/leds/0" → segment(2) = "0").
    pub fn segment(self: Request, idx: i32) ?[*:0]const u8 {
        const p = c.ove_httpd_req_segment(self.raw, idx);
        if (p == null) return null;
        return @ptrCast(p);
    }
};

/// HTTP response writer (opaque, borrowed for handler duration).
pub const ResponseWriter = struct {
    raw: *c.ove_httpd_resp_t,

    /// Send a JSON response.
    pub fn json(self: ResponseWriter, status_code: i32, json_str: [:0]const u8) Error!void {
        try err.fromCode(c.ove_httpd_resp_json(self.raw, status_code, json_str.ptr));
    }

    /// Send an HTML response.
    pub fn html(self: ResponseWriter, status_code: i32, html_str: []const u8) Error!void {
        try err.fromCode(c.ove_httpd_resp_html(self.raw, status_code, html_str.ptr, html_str.len));
    }

    /// Send a response with explicit content type.
    pub fn send(self: ResponseWriter, status_code: i32, content_type: [:0]const u8, body_data: []const u8) Error!void {
        try err.fromCode(c.ove_httpd_resp_send(self.raw, status_code, content_type.ptr, body_data.ptr, body_data.len));
    }

    /// Send a pre-gzipped response (adds Content-Encoding: gzip).
    pub fn sendGz(self: ResponseWriter, status_code: i32, content_type: [:0]const u8, body_data: []const u8) Error!void {
        try err.fromCode(c.ove_httpd_resp_send_gz(self.raw, status_code, content_type.ptr, body_data.ptr, body_data.len));
    }

    /// Send an error response.
    pub fn errResponse(self: ResponseWriter, status_code: i32, message: [:0]const u8) Error!void {
        try err.fromCode(c.ove_httpd_resp_error(self.raw, status_code, message.ptr));
    }
};

// ---------------------------------------------------------------------------
// Server functions
// ---------------------------------------------------------------------------

/// Server configuration.
pub const ServerConfig = struct {
    port: u16 = 80,
    max_body_size: i32 = 1024,
};

/// Start the HTTP server.
pub fn start(cfg: ServerConfig) Error!void {
    var cc: c.ove_httpd_config_t = std.mem.zeroes(c.ove_httpd_config_t);
    cc.port = cfg.port;
    cc.max_body_size = cfg.max_body_size;
    try err.fromCode(c.ove_httpd_start(&cc));
}

/// Stop the HTTP server.
pub fn stop() void {
    c.ove_httpd_stop();
}

/// Register a route with a typed Zig handler.
///
/// The handler is wrapped in a comptime trampoline that converts the
/// raw C pointers into `Request` and `ResponseWriter` values.
pub fn route(
    method_str: [:0]const u8,
    path_str: [:0]const u8,
    comptime handler: fn (Request, ResponseWriter) Error!void,
) Error!void {
    const Trampoline = struct {
        fn invoke(req: ?*c.ove_httpd_req_t, resp: ?*c.ove_httpd_resp_t) callconv(.c) c_int {
            const zig_req = Request{ .raw = req orelse return c.OVE_ERR_INVALID_PARAM };
            const zig_resp = ResponseWriter{ .raw = resp orelse return c.OVE_ERR_INVALID_PARAM };
            handler(zig_req, zig_resp) catch |e| {
                return switch (e) {
                    error.InvalidParam => c.OVE_ERR_INVALID_PARAM,
                    error.NoMemory => c.OVE_ERR_NO_MEMORY,
                    error.NotSupported => c.OVE_ERR_NOT_SUPPORTED,
                    else => -1,
                };
            };
            return 0;
        }
    };
    try err.fromCode(c.ove_httpd_route(method_str.ptr, path_str.ptr, &Trampoline.invoke));
}

/// Register the built-in Tasmota-style dashboard routes.
pub fn registerBuiltinRoutes() void {
    c.ove_httpd_register_builtin_routes();
}

/// Set the network interface used by built-in dashboard routes.
pub fn setNetif(handle: c.ove_netif_t) void {
    c.ove_httpd_set_netif(handle);
}

/// Append a log line to the httpd log ring buffer.
pub fn logAppend(line: [:0]const u8) void {
    c.ove_httpd_log_append(line.ptr);
}

// ---------------------------------------------------------------------------
// WebSocket
// ---------------------------------------------------------------------------

/// WebSocket support (available when CONFIG_OVE_NET_HTTPD_WS is enabled).
pub const ws = if (@hasDecl(c, "ove_httpd_ws_route")) struct {
    /// Register a WebSocket route.
    ///
    /// `on_message` and `on_close` may be null.
    pub fn route(
        path_str: [:0]const u8,
        on_message: c.ove_httpd_ws_handler_t,
        on_close: c.ove_httpd_ws_close_handler_t,
    ) Error!void {
        try err.fromCode(c.ove_httpd_ws_route(path_str.ptr, on_message, on_close));
    }

    /// Send a message to a WebSocket connection.
    pub fn send(conn: *c.ove_httpd_ws_conn_t, data: []const u8) Error!void {
        try err.fromCode(c.ove_httpd_ws_send(conn, data.ptr, data.len));
    }

    /// Broadcast a message to all WebSocket connections on a path.
    ///
    /// Returns the number of connections the message was sent to.
    pub fn broadcast(path_str: [:0]const u8, data: []const u8) c_int {
        return c.ove_httpd_ws_broadcast(path_str.ptr, data.ptr, data.len);
    }

    /// Return the number of active WebSocket connections.
    pub fn activeCount() c_int {
        return c.ove_httpd_ws_active_count();
    }
} else struct {};

File: bindings/rust/ove/src/net_httpd.rs
// Copyright (C) 2026 Kamil Lulko <kamil.lulko@gmail.com>
//
// SPDX-License-Identifier: GPL-3.0-or-later
//
// This file is part of oveRTOS.

//! Embedded HTTP server with REST-style routing.
//!
//! Wraps the oveRTOS `ove_httpd_*` API.  The server runs on a background
//! task; routes are registered with [`route`] and dispatched to
//! [`Handler`] callbacks that receive a [`Request`] / [`Response`] pair.
//!
//! The server is a singleton (one per process), so start/stop/route are
//! free functions rather than methods on a struct.

use core::fmt;
use core::marker::PhantomData;

use crate::bindings;
use crate::error::{Error, Result};

// ---------------------------------------------------------------------------
// Handler type
// ---------------------------------------------------------------------------

/// Route handler callback.
///
/// This is the raw C function pointer type:
/// `fn(req: *mut ove_httpd_req_t, resp: *mut ove_httpd_resp_t) -> i32`.
pub type Handler = bindings::ove_httpd_handler_t;

// ---------------------------------------------------------------------------
// Request
// ---------------------------------------------------------------------------

/// HTTP request passed to a route handler.
///
/// Borrows the underlying C request for the duration of the handler call.
/// Do not store this value beyond the handler invocation.
pub struct Request<'a> {
    raw: *mut bindings::ove_httpd_req_t,
    _marker: PhantomData<&'a ()>,
}

impl<'a> Request<'a> {
    /// Create a `Request` from a raw C pointer.
    ///
    /// # Safety
    /// `raw` must be a valid pointer obtained from an active httpd handler
    /// callback and must remain valid for the lifetime `'a`.
    pub unsafe fn from_raw(raw: *mut bindings::ove_httpd_req_t) -> Self {
        Self {
            raw,
            _marker: PhantomData,
        }
    }

    /// The HTTP method string (e.g. "GET", "POST").
    pub fn method(&self) -> &[u8] {
        let p = unsafe { bindings::ove_httpd_req_method(self.raw) };
        if p.is_null() {
            return &[];
        }
        unsafe { core::ffi::CStr::from_ptr(p).to_bytes() }
    }

    /// The full request path (e.g. "/api/leds/0").
    pub fn path(&self) -> &[u8] {
        let p = unsafe { bindings::ove_httpd_req_path(self.raw) };
        if p.is_null() {
            return &[];
        }
        unsafe { core::ffi::CStr::from_ptr(p).to_bytes() }
    }

    /// The query string after '?' (empty slice if none).
    pub fn query(&self) -> &[u8] {
        let p = unsafe { bindings::ove_httpd_req_query(self.raw) };
        if p.is_null() {
            return &[];
        }
        unsafe { core::ffi::CStr::from_ptr(p).to_bytes() }
    }

    /// The request body (empty slice if none).
    pub fn body(&self) -> &[u8] {
        let p = unsafe { bindings::ove_httpd_req_body(self.raw) };
        let len = unsafe { bindings::ove_httpd_req_body_len(self.raw) };
        if p.is_null() || len == 0 {
            return &[];
        }
        unsafe { core::slice::from_raw_parts(p as *const u8, len) }
    }

    /// A path segment by index.
    ///
    /// For path "/api/leds/0": segment 0 = "api", 1 = "leds", 2 = "0".
    /// Returns `None` if the index is out of range.
    pub fn segment(&self, idx: i32) -> Option<&[u8]> {
        let p = unsafe { bindings::ove_httpd_req_segment(self.raw, idx) };
        if p.is_null() {
            return None;
        }
        Some(unsafe { core::ffi::CStr::from_ptr(p).to_bytes() })
    }
}

impl fmt::Debug for Request<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Request")
            .field(
                "path",
                &core::str::from_utf8(self.path()).unwrap_or("<non-utf8>"),
            )
            .finish()
    }
}

// ---------------------------------------------------------------------------
// Response
// ---------------------------------------------------------------------------

/// HTTP response writer passed to a route handler.
///
/// Borrows the underlying C response for the duration of the handler call.
/// Do not store this value beyond the handler invocation.
pub struct Response<'a> {
    raw: *mut bindings::ove_httpd_resp_t,
    _marker: PhantomData<&'a ()>,
}

impl<'a> Response<'a> {
    /// Create a `Response` from a raw C pointer.
    ///
    /// # Safety
    /// `raw` must be a valid pointer obtained from an active httpd handler
    /// callback and must remain valid for the lifetime `'a`.
    pub unsafe fn from_raw(raw: *mut bindings::ove_httpd_resp_t) -> Self {
        Self {
            raw,
            _marker: PhantomData,
        }
    }

    /// Send a JSON response.
    ///
    /// `json` must be a null-terminated byte string.
    ///
    /// # Errors
    /// Returns an error if the response cannot be sent.
    pub fn json(&self, status: i32, json: &[u8]) -> Result<()> {
        let rc =
            unsafe { bindings::ove_httpd_resp_json(self.raw, status, json.as_ptr() as *const _) };
        Error::from_code(rc)
    }

    /// Send an HTML response.
    ///
    /// # Errors
    /// Returns an error if the response cannot be sent.
    pub fn html(&self, status: i32, html: &[u8]) -> Result<()> {
        let rc = unsafe {
            bindings::ove_httpd_resp_html(self.raw, status, html.as_ptr() as *const _, html.len())
        };
        Error::from_code(rc)
    }

    /// Send a response with arbitrary content type.
    ///
    /// `content_type` must be a null-terminated byte string.
    ///
    /// # Errors
    /// Returns an error if the response cannot be sent.
    pub fn send(&self, status: i32, content_type: &[u8], body: &[u8]) -> Result<()> {
        let rc = unsafe {
            bindings::ove_httpd_resp_send(
                self.raw,
                status,
                content_type.as_ptr() as *const _,
                body.as_ptr() as *const _,
                body.len(),
            )
        };
        Error::from_code(rc)
    }

    /// Send a JSON error response.
    ///
    /// `msg` must be a null-terminated byte string.
    ///
    /// # Errors
    /// Returns an error if the response cannot be sent.
    pub fn error(&self, status: i32, msg: &[u8]) -> Result<()> {
        let rc =
            unsafe { bindings::ove_httpd_resp_error(self.raw, status, msg.as_ptr() as *const _) };
        Error::from_code(rc)
    }

    /// Send a pre-gzipped response (adds `Content-Encoding: gzip`).
    ///
    /// `content_type` must be a null-terminated byte string.
    ///
    /// # Errors
    /// Returns an error if the response cannot be sent.
    pub fn send_gz(&self, status: i32, content_type: &[u8], body: &[u8]) -> Result<()> {
        let rc = unsafe {
            bindings::ove_httpd_resp_send_gz(
                self.raw,
                status,
                content_type.as_ptr() as *const _,
                body.as_ptr() as *const _,
                body.len(),
            )
        };
        Error::from_code(rc)
    }
}

impl fmt::Debug for Response<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Response")
            .field("raw", &format_args!("{:p}", self.raw))
            .finish()
    }
}

// ---------------------------------------------------------------------------
// Server free functions
// ---------------------------------------------------------------------------

/// Start the HTTP server on the given port.
///
/// Spawns a background task that accepts connections.  Register routes
/// with [`route`] before or after starting.
///
/// # Errors
/// Returns an error if the server cannot bind or start.
pub fn start(port: u16, max_body: i32) -> Result<()> {
    let cfg = bindings::ove_httpd_config_t {
        port,
        max_body_size: max_body,
    };
    let rc = unsafe { bindings::ove_httpd_start(&cfg) };
    Error::from_code(rc)
}

/// Stop the HTTP server and close the listening socket.
pub fn stop() {
    unsafe { bindings::ove_httpd_stop() }
}

/// Register a route handler.
///
/// `method` and `path` must be null-terminated byte strings
/// (e.g. `b"GET\0"`, `b"/api/leds\0"`).
///
/// # Errors
/// Returns an error if the route table is full.
pub fn route(method: &[u8], path: &[u8], handler: Handler) -> Result<()> {
    let rc = unsafe {
        bindings::ove_httpd_route(
            method.as_ptr() as *const _,
            path.as_ptr() as *const _,
            handler,
        )
    };
    Error::from_code(rc)
}

/// Register the built-in dashboard routes (`/api/info`, `/api/leds`, etc.).
///
/// Call after [`start`] to add the standard device management API.
pub fn register_builtin_routes() {
    unsafe { bindings::ove_httpd_register_builtin_routes() }
}

/// Append a log line to the httpd log ring buffer.
///
/// Call from a log output hook to capture lines for `GET /api/log`.
pub fn log_append(line: &[u8]) {
    unsafe { bindings::ove_httpd_log_append(line.as_ptr() as *const _) }
}

/// Bind the HTTP server to a specific network interface.
pub fn set_netif(handle: bindings::ove_netif_t) {
    unsafe { bindings::ove_httpd_set_netif(handle) }
}

// ---------------------------------------------------------------------------
// WebSocket support
// ---------------------------------------------------------------------------

#[cfg(has_net_httpd_ws)]
pub mod ws {
    use crate::bindings;
    use crate::error::{Error, Result};

    /// Register a WebSocket route.
    ///
    /// `path` must be a null-terminated byte string (e.g. `b"/ws\0"`).
    ///
    /// # Errors
    /// Returns an error if the route table is full.
    pub fn route(
        path: &[u8],
        on_message: bindings::ove_httpd_ws_handler_t,
        on_close: bindings::ove_httpd_ws_close_handler_t,
    ) -> Result<()> {
        let rc = unsafe {
            bindings::ove_httpd_ws_route(path.as_ptr() as *const _, on_message, on_close)
        };
        Error::from_code(rc)
    }

    /// Send a text message to a WebSocket connection.
    ///
    /// # Errors
    /// Returns an error if the send fails.
    pub fn send(conn: *mut bindings::ove_httpd_ws_conn_t, data: &[u8]) -> Result<()> {
        let rc =
            unsafe { bindings::ove_httpd_ws_send(conn, data.as_ptr() as *const _, data.len()) };
        Error::from_code(rc)
    }

    /// Broadcast a message to all WebSocket connections on a path.
    ///
    /// `path` must be a null-terminated byte string.
    ///
    /// Returns the number of connections the message was sent to.
    pub fn broadcast(path: &[u8], data: &[u8]) -> i32 {
        unsafe {
            bindings::ove_httpd_ws_broadcast(
                path.as_ptr() as *const _,
                data.as_ptr() as *const _,
                data.len(),
            )
        }
    }

    /// Return the number of active WebSocket connections.
    pub fn active_count() -> i32 {
        unsafe { bindings::ove_httpd_ws_active_count() }
    }
}
